### ISSUE TITLE
validate_patches.py errors are too vague

### ISSUE THREAD

**Issue by braewoods** [Eloston/ungoogled-chromium#957]
I have been fixing patch validation issues in downstream repos for awhile now and I have observed that the validator works fine but the errors it gives even in verbose mode are next to useless for finding out what the issue is. I often have to resort to using the regular patch program just to find out what the issue is as the information from a dry run of the patch program can offer me precise information about how the patch offsets need to be adjusted to make the validator happy or even about a patch that needs to be dropped entirely because it is now redundant.

Is there any way to improve this situation?

**Comment by Eloston** [comment]
`validate_patches.py` uses a custom in-memory patching algorithm to validate patches, so we can't easily add more verbosity to the Python implementation.

However, if `patch --dry-run` output is sufficient, then we can modify `validate_patches.py` to do a dry run on the patch when `-v` is given. A few notes about the implementation:

* GNU patch requires the files to exist on the filesystem. Therefore, we could temporarily save the in-memory cache of patched files (that are required by the patch) to the filesystem (via module `tempfile`), and then let `patch` do a dry run on these files.
* We should separate the logic to find the patch executable from `patches.apply_patches`. Then we could add a new method `patches.dry_run_check` to capture the stdout/stderr from running `patch --dry-run`

Do you think this would work?

**Comment by braewoods** [comment]
Probably but there is another thing to consider. How portable would it be? Patch is primarily a unix utility so would it work on Windows at all?

**Comment by Eloston** [comment]
@braewoods https://github.com/Eloston/ungoogled-chromium/blob/master/utils/patches.py#L31-L35

tl;dr Yes

Also, we should move that Windows-specific logic to the Windows repo...
